### ISSUE TITLE
[Feature request] There is no converter API from mx.nd.NDArray to DLPack tensor

### ISSUE THREAD

**Issue by wkcn** [apache/incubator-mxnet#11964]
Hi, all.

[DLPack](https://github.com/dmlc/dlpack) is important for sharing tensors and operators among Deep Learning Frameworks.

PyTorch provides the API `torch.utils.dlpack.to_dlpack` and `torch.utils.dlpack.from_dlpack`.

MXNet also uses DLPack, however it seems that there is no converter API from mx.nd.NDArray to DLPack tensor ([DLManagedTensor](https://github.com/dmlc/dlpack/blob/master/include/dlpack/dlpack.h#L125)).

I think it's better to add the API.

Thank you!

---------------------------------
Note:

In [TVM](https://github.com/dmlc/tvm),
[DLManagedTensor* NDArray::ToDLPack() const](https://github.com/dmlc/tvm/blob/master/src/runtime/ndarray.cc#L117)
[NDArray NDArray::FromDLPack(DLManagedTensor* tensor)](https://github.com/dmlc/tvm/blob/master/src/runtime/ndarray.cc#L140)

However, it seems that `NDArray` of MXNet and TVM are different.

**Comment by Roshrini** [comment]
@sandeep-krishnamurthy Can you please add label: FeatureRequest

**Comment by arogozhnikov** [comment]
`cupy` (version > 5) also supports dlpack.

**Comment by tqchen** [comment]
This is a good proposal, there is one thing that we need to be cautious about: MXNet uses asynchronous execution and things can run out of order if it is not tracked by mxnet engine.

Here is an example that things can go wrong.
```python
x = mxnet.nd.zeros((3,4))
x = x + 1
# pass to pytorch or tvm
y = other_framework.nd.fromdlpack(x.asdlpack())
# Y may or maynot be 1
print(y)
```
This is because when we directly call ```x.asdlpack()``` and directly convert the data structure, the execution may not yet finished.

### A Fix
We need to also add flags to asdlpack

```python
x = mxnet.nd.zeros((3,4))
x = x + 1
# pass to pytorch or tvm
x.wait_for_read()
y = other_framework.nd.fromdlpack(x.asdlpack)
# Y may or maynot be 1
print(y)
```
This works for read case, similarly, if we plan to modify the content, we need to do
```python
x = mxnet.nd.zeros((3,4))
x = x + 1
# pass to pytorch or tvm
x.wait_for_write()
y = other_framework.nd.fromdlpack(x.asdlpack)
y = y + 2
# Y should be 3
print(y)
```

### API design
To make sure things won't go wrong because the user does not understand things, I would recommend we use the following API

- NDArray.asdlpack_for_read()
- NDArray.asdlpack_for_write()

**Comment by tqchen** [comment]
In terms of API naming, I would recommend avoid using asdplack, the main reason is that most asxxx API in numpy implies copy of data. For example, astype, asarray, ascontiguousarray all would imply possible data copy.

While we are creating a memory view here, I think to_dlpack is a better name

**Comment by wkcn** [comment]
@tqchen Thank you!

**Comment by tqchen** [comment]
close by #12047

**Comment by tqchen** [comment]
Thanks @wkcn this is now part of MXNet, do you mind to do a write up on introducing this new feature?

**Comment by wkcn** [comment]
@tqchen I'm sorry that I have some important thing to do and little time to write it recently.
